src/devices/uart.c

```diff
@@ -0,0 +1,160 @@
+
+    /* Update current interrupt (higher bits -> more priority) */
+    if (uart->pending_intrs)
+        uart->current_intr = ilog2(uart->pending_intrs);
+}
+
+void u8250_check_ready(u8250_state_t *uart)
+{
+    if (uart->in_ready)
+        return;
+
+    struct pollfd pfd = {uart->in_fd, POLLIN, 0};
+    poll(&pfd, 1, 0);
+    if (pfd.revents & POLLIN)
+        uart->in_ready = true;
+}
+
+static void u8250_handle_out(u8250_state_t *uart, uint8_t value)
+{
+    if (write(uart->out_fd, &value, 1) < 1)
+        fprintf(stderr, "failed to write UART output: %s\n", strerror(errno));
+}
+
+static uint8_t u8250_handle_in(u8250_state_t *uart)
+{
+    uint8_t value = 0;
+    u8250_check_ready(uart);
+    if (!uart->in_ready)
+        return value;
+
+    if (read(uart->in_fd, &value, 1) < 0)
+        fprintf(stderr, "failed to read UART input: %s\n", strerror(errno));
+    uart->in_ready = false;
+    u8250_check_ready(uart);
+
+    if (value == 1) {           /* start of heading (Ctrl-a) */
+        if (getchar() == 120) { /* keyboard x */
+            printf("\n");       /* end emulator with newline */
+            exit(0);
+        }
+    }
+
+    return value;
+}
+
+static uint8_t u8250_reg_read(u8250_state_t *uart, uint32_t addr)
+{
+    uint8_t ret;
+    switch (addr) {
+    case 0:
+        if (uart->lcr & (1 << 7)) /* DLAB */
+            return uart->dll;
+        return u8250_handle_in(uart);
+    case 1:
+        if (uart->lcr & (1 << 7)) /* DLAB */
+            return uart->dlh;
+        return uart->ier;
+    case 2:
+        ret = (uart->current_intr << 1) | (uart->pending_intrs ? 0 : 1);
+        if (uart->current_intr == U8250_INT_THRE)
+            uart->pending_intrs &= ~(1 << uart->current_intr);
+        return ret;
+    case 3:
+        return uart->lcr;
+    case 4:
+        return uart->mcr;
+        break;
+    case 5:
+        /* LSR = no error, TX done & ready */
+        return (0x60 | (uint8_t) uart->in_ready);
+    case 6:
+        /* MSR = carrier detect, no ring, data ready, clear to send. */
+        return 0xb0;
+        /* no scratch register, so we should be detected as a plain 8250. */
+    default:
+        return 0;
+    }
+
+    return 0;
+}
+
+static void u8250_reg_write(u8250_state_t *uart, uint32_t addr, uint8_t value)
+{
+    switch (addr) {
+    case 0:
+        if (uart->lcr & (1 << 7)) { /* DLAB */
+            uart->dll = value;
+            break;
+        }
+        u8250_handle_out(uart, value);
+        uart->pending_intrs |= 1 << U8250_INT_THRE;
+        break;
+    case 1:
+        if (uart->lcr & (1 << 7)) { /* DLAB */
+            uart->dlh = value;
+            break;
+        }
+        uart->ier = value;
+        break;
+    case 3:
+        uart->lcr = value;
+        break;
+    case 4:
+        uart->mcr = value;
+        break;
+    }
+}
+
+uint32_t u8250_read(u8250_state_t *uart, uint32_t addr)
+{
+    return (uint32_t) (int8_t) u8250_reg_read(uart, addr);
```

Comment:
I think we are good to inline `u8250_reg_read()` and `u8250_reg_write()` manually.

Reply:
How about using `FORCE_INLINE`(defined in [here](https://github.com/sysprog21/rv32emu/blob/451f8c0a9990e9beb4be8c286c07f67ed1bbbf81/src/common.h#L14)) macro?

Reply:
No. What I'm considering is, if we never reuse `u8250_reg_read()` and `u8250_reg_write()`, I don't see any advantage to implementing them like this.